src/dsp/rde.c

```diff
@@ -1002,3 +1002,198 @@ int decode_rde_multipart_send_resp(const struct pldm_msg *msg,
+		return rc;
+	}
+
+	pldm_msgbuf_extract_p(buf, data_transfer_handle);
+	pldm_msgbuf_extract_p(buf, operation_id);
+	pldm_msgbuf_extract_p(buf, transfer_operation);
+
+	if (*transfer_operation > PLDM_RDE_XFER_ABORT) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+
+	return pldm_msgbuf_complete(buf);
+}
+
+LIBPLDM_ABI_STABLE
+int encode_rde_multipart_receive_resp(
+	uint8_t instance_id, uint8_t completion_code, uint8_t transfer_flag,
+	uint32_t next_data_transfer_handle, uint32_t data_length_bytes,
+	bool add_checksum, uint32_t data_integrity_checksum,
+	const uint8_t *data, struct pldm_msg *msg)
+{
+	PLDM_MSGBUF_DEFINE_P(buf);
+	int rc;
+
+	if (NULL == msg) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+
+	struct pldm_header_info header;
+	header.msg_type = PLDM_RESPONSE;
+	header.instance = instance_id;
+	header.pldm_type = PLDM_RDE;
+	header.command = PLDM_RDE_MULTIPART_RECEIVE;
+	rc = pack_pldm_header(&header, &(msg->hdr));
+
+	if (rc != PLDM_SUCCESS) {
+		return rc;
+	}
+
+	if (completion_code != PLDM_SUCCESS) {
+		msg->payload[0] = completion_code;
+		return PLDM_SUCCESS;
+	}
+
+	rc = pldm_msgbuf_init_errno(
+		buf, PLDM_RDE_MULTIPART_RECEIVE_RESP_FIXED_BYTES, msg->payload,
+		PLDM_RDE_MULTIPART_RECEIVE_RESP_FIXED_BYTES +
+			data_length_bytes);
+	if (rc) {
+		return rc;
+	}
+
+	pldm_msgbuf_insert(buf, completion_code);
+	pldm_msgbuf_insert(buf, transfer_flag);
+	pldm_msgbuf_insert(buf, next_data_transfer_handle);
+	pldm_msgbuf_insert(buf, data_length_bytes);
+
+	rc = pldm_msgbuf_insert_array(
+		buf,
+		data_length_bytes -
+			(add_checksum ? sizeof(data_integrity_checksum) : 0),
+		(const uint8_t *)data, data_length_bytes);
+	if (rc != PLDM_SUCCESS) {
+		return rc;
+	}
+
+	if (add_checksum) {
+		pldm_msgbuf_insert(buf, data_integrity_checksum);
+	}
+
+	return pldm_msgbuf_complete(buf);
+}
+
+LIBPLDM_ABI_STABLE
+int decode_rde_multipart_receive_resp(
+	const struct pldm_msg *msg, size_t payload_length,
+	uint8_t *completion_code, uint8_t *transfer_flag,
+	uint32_t *data_transfer_handle, uint32_t *data_length_bytes,
+	uint8_t *data, uint32_t *data_integrity_checksum)
+{
+	PLDM_MSGBUF_DEFINE_P(buf);
+	int rc;
+	bool add_checksum = false;
+
+	if ((msg == NULL) || (completion_code == NULL) ||
+	    (transfer_flag == NULL) || (data_transfer_handle == NULL) ||
+	    (data_length_bytes == NULL) || (data == NULL)) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+
+	*completion_code = msg->payload[0];
+
+	if (PLDM_SUCCESS != *completion_code) {
+		fprintf(stderr,
+			"Unsuccessful completion code received in multipart receive: %u\n",
+			(uint8_t)(*completion_code));
+		return PLDM_ERROR;
+	}
+
+	if (payload_length < PLDM_RDE_MULTIPART_RECEIVE_RESP_FIXED_BYTES) {
+		return PLDM_ERROR_INVALID_LENGTH;
+	}
+	rc = pldm_msgbuf_init_errno(buf,
+				    PLDM_RDE_MULTIPART_RECEIVE_RESP_FIXED_BYTES,
+				    msg->payload, payload_length);
+	if (rc != PLDM_SUCCESS) {
+		fprintf(stderr, "init failed\n");
+		return rc;
+	}
+
+	pldm_msgbuf_extract_p(buf, completion_code);
```

Comment:
check completeion code value , and act based on spec. like no point in collecting remaing data for error case , this is applicalble all decode function.

CompletionCode
value: { PLDM_BASE_CODES, ERROR_OPERATION_ABANDONED,
ERROR_OPERATION_FAILED, ERROR_UNEXPECTED, ERROR_BAD_CHECKSUM }
If the DataTransferHandle does not correspond to a valid chunk, the RDE Device shall return
CompletionCode ERROR_INVALID_DATA.
If the transfer is aborted, the RDE Device shall acknowledge this status by returning SUCCESS.